### PR TITLE
[TS migration] Prepare remaining files for Onyx.js migration

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details
This PR fixes various types across Onyx, also adjusting tests in order to fix failing typecheck. 

### Related Issues
Part of https://github.com/Expensify/App/issues/34344

### Automated Tests
N/A

### Manual Tests
N/A

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos

Code is only changed on the type level, I verified it works on web, no further tests needed.

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/react-native-onyx/assets/46095609/b99932b9-c17e-4a2a-946d-48410a876dee

</details>

### CODE REVIEW

File: lib/storage/__mocks__/index.ts [hunk 1]
```diff
@@ -34,8 +35,7 @@ const idbKeyvalMock: StorageProvider = {
     multiMerge(pairs) {
         pairs.forEach(([key, value]) => {
             const existingValue = storageMapInternal[key];
-            // eslint-disable-next-line @typescript-eslint/no-explicit-any
-            const newValue = utils.fastMerge(existingValue as any, value);
+            const newValue = utils.fastMerge(existingValue!, value!);
```
Comment (fabioh8010) [inline lib/storage/__mocks__/index.ts:38]: Is this really safe?

File: lib/storage/providers/IDBKeyVal.ts [hunk 2]
```diff
@@ -21,13 +21,12 @@ const provider: StorageProvider = {
...
             // of the items in one transaction to achieve best performance.
-            const getValues = Promise.all(pairs.map(([key]) => promisifyRequest<Value>(store.get(key))));
+            const getValues = Promise.all(pairs.map(([key]) => promisifyRequest<OnyxValue>(store.get(key))));
 
             return getValues.then((values) => {
                 const upsertMany = pairs.map(([key, value], index) => {
                     const prev = values[index];
-                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                    const newValue = utils.fastMerge(prev as any, value);
+                    const newValue = utils.fastMerge(prev!, value!);
```
Comment (fabioh8010) [inline lib/storage/providers/IDBKeyVal.ts:29]: Same

File: lib/types.ts [hunk 3]
```diff
@@ -102,7 +103,7 @@ type OnyxKey = Key | CollectionKey;
 /**
  * Represents a Onyx value that can be either a single entry or a collection of entries, depending on the `TKey` provided.
  */
-type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
+type OnyxValue<TKey extends OnyxKey = OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
```
Comment (fabioh8010) [inline lib/types.ts:106]: Why is this change needed?

File: tests/unit/useOnyxTest.ts [hunk 4]
```diff
@@ -98,7 +98,7 @@ describe('useOnyx', () => {
         });
 
         it('should initially return null while loading non-cached key, and then return value and loaded state', async () => {
-            await StorageMock.setItem(ONYXKEYS.TEST_KEY, 'test');
+            await StorageMock.setItem(ONYXKEYS.TEST_KEY, {test: 'test'});
```
Comment (fabioh8010) [inline tests/unit/useOnyxTest.ts:101]: Why these changes in this file?

Comment (blazejkustra) [reply tests/unit/useOnyxTest.ts:101]: Typecheck is complaining in this file due to `OnyxValue` implementation:
```ts
type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
```

second argument of `StorageMock.setItem` results in `OnyxCollection<unknown>`, that's why I changed it here. Any other ideas?

File: lib/types.ts (hunk 3 above)
Comment (blazejkustra) [reply lib/types.ts:106]: When I look at it now, I agree this should not have a default value. This will require some changes 👍

File: lib/storage/__mocks__/index.ts (hunk 1 above)
Comment (blazejkustra) [reply lib/storage/__mocks__/index.ts:38]: I changed the typings, should be better now 😄

File: lib/utils.ts [hunk 5]
```diff
@@ -26,15 +26,15 @@ function isMergeableObject(value: unknown): boolean {
  * @param shouldRemoveNullObjectValues - If true, null object values will be removed.
  * @returns - The merged object.
  */
-function mergeObject<TTarget extends unknown[] | Record<string, unknown>>(target: TTarget, source: TTarget, shouldRemoveNullObjectValues = true): TTarget {
+function mergeObject<TValue extends OnyxValue<OnyxKey>>(target: OnyxEntry<TValue>, source: TValue, shouldRemoveNullObjectValues = true): TValue {
```
Comment (fabioh8010) [inline lib/utils.ts:29]: Should we restrict this function to only onyx values? I think it was supposed to be used as a generic utils

File: lib/utils.ts [hunk 6]
```diff
@@ -80,7 +80,7 @@ function mergeObject<TTarget extends unknown[] | Record<string, unknown>>(target
  * On native, when merging an existing value with new changes, SQLite will use JSON_PATCH, which removes top-level nullish values.
  * To be consistent with the behaviour for merge, we'll also want to remove null values for "set" operations.
  */
-function fastMerge<TTarget extends unknown[] | Record<string, unknown>>(target: TTarget, source: TTarget, shouldRemoveNullObjectValues = true): TTarget {
+function fastMerge<TValue extends OnyxValue<OnyxKey>>(target: TValue, source: TValue, shouldRemoveNullObjectValues = true): TValue {
```
Comment (fabioh8010) [inline lib/utils.ts:83]: Same

File: tests/unit/useOnyxTest.ts (hunk 4 above)
Comment (fabioh8010) [reply tests/unit/useOnyxTest.ts:101]: I'm in favor of either trying to find a fix for `OnyxCollection<unknown>` case or using `// @ts-expect-error` because otherwise it would restrict our tests

File: tests/unit/useOnyxTest.ts [hunk 7]
```diff
@@ -98,7 +98,7 @@ describe('useOnyx', () => {
         });
 
         it('should initially return null while loading non-cached key, and then return value and loaded state', async () => {
-            await StorageMock.setItem(ONYXKEYS.TEST_KEY, {test: 'test'});
+            await StorageMock.setItem(ONYXKEYS.TEST_KEY, 'test' as unknown);
```
Comment (fabioh8010) [inline tests/unit/useOnyxTest.ts:101]:
```suggestion
            await StorageMock.setItem(ONYXKEYS.TEST_KEY, 'test');
```

Do we need the assertion?

**Review by fabioh8010 — APPROVED** [review]
🚀

**Review by MariaHCD — APPROVED** [review]
